ceno_zkvm/src/e2e.rs

```diff
@@ -6,135 +6,64 @@ use crate::{
     },
     state::GlobalState,
     structs::{ProgramParams, ZKVMConstraintSystem, ZKVMFixedTraces, ZKVMWitnesses},
-    tables::{MemFinalRecord, MemInitRecord, ProgramTableCircuit},
+    tables::{MemFinalRecord, MemInitRecord, ProgramTableCircuit, ProgramTableConfig},
 };
 use ceno_emul::{
     ByteAddr, EmuContext, InsnKind::EANY, IterAddresses, Platform, Program, StepRecord, Tracer,
     VMState, WORD_SIZE, WordAddr,
 };
 use ff_ext::ExtensionField;
-use itertools::{Itertools, MinMaxResult, chain, enumerate};
+use itertools::{Itertools, MinMaxResult, chain};
 use mpcs::PolynomialCommitmentScheme;
 use std::{
     collections::{HashMap, HashSet},
     iter::zip,
-    time::Instant,
 };
 use transcript::Transcript;
 
-type E2EWitnessGen<E, PCS> = (
-    ZKVMProver<E, PCS>,
-    ZKVMVerifier<E, PCS>,
-    ZKVMWitnesses<E>,
-    PublicValues<u32>,
-    usize,   // number of cycles
-    Instant, // e2e start, excluding key gen time
-    Option<u32>,
-);
-
-pub fn run_e2e_gen_witness<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>>(
-    program: Program,
-    platform: Platform,
-    stack_size: u32,
-    heap_size: u32,
-    hints: Vec<u32>,
-    max_steps: usize,
-) -> E2EWitnessGen<E, PCS> {
-    let stack_addrs = platform.stack_top - stack_size..platform.stack_top;
-
-    // Detect heap as starting after program data.
-    let heap_start = program.image.keys().max().unwrap() + WORD_SIZE as u32;
-    let heap_addrs = heap_start..heap_start + heap_size;
-
-    let mut mem_padder = MemPadder::new(heap_addrs.end..platform.ram.end);
-
-    let mem_init = {
-        let program_addrs = program.image.iter().map(|(addr, value)| MemInitRecord {
-            addr: *addr,
-            value: *value,
-        });
-
-        let stack = stack_addrs
-            .iter_addresses()
-            .map(|addr| MemInitRecord { addr, value: 0 });
+struct FullMemState<Record> {
+    mem: Vec<Record>,
+    io: Vec<Record>,
+    reg: Vec<Record>,
+    priv_io: Vec<Record>,
+}
 
-        let heap = heap_addrs
-            .iter_addresses()
-            .map(|addr| MemInitRecord { addr, value: 0 });
+type InitMemState = FullMemState<MemInitRecord>;
+type FinalMemState = FullMemState<MemFinalRecord>;
 
-        let mem_init = chain!(program_addrs, stack, heap).collect_vec();
+struct EmulationResult {
+    exit_code: Option<u32>,
+    all_records: Vec<StepRecord>,
+    final_mem_state: FinalMemState,
+    pi: PublicValues<u32>,
+}
 
-        mem_padder.padded_sorted(mem_init.len().next_power_of_two(), mem_init)
-    };
+fn emulate_program(
+    program: &Program,
+    max_steps: usize,
+    init_mem_state: InitMemState,
+    platform: &Platform,
+    hints: Vec<u32>,
+) -> EmulationResult {
+    let InitMemState {
+        mem: mem_init,
+        io: io_init,
+        reg: reg_init,
+        priv_io: _,
+    } = init_mem_state;
 
-    let mut vm = VMState::new(platform.clone(), program);
+    let mut vm: VMState = VMState::new(platform.clone(), program.clone());
```

Comment:
program clone here might be an overhead for e2e. probably
- plan a: change VMState to new with program reference, which need to introduce new life cycle.
- plan b or we can reuse same VMState with a reset() function

Plan a seems bring less change comparing with plan b

Reply:
`VMState` just wanted an `Arc` which I propagated to the top level. Clones are cheap now.